compiler-rt/lib/rtsan/rtsan_checks.inc

```diff
@@ -0,0 +1,20 @@
+//===-- rtsan_checks.inc ----------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// List of checks handled by RTSan runtime.
+//
+//===----------------------------------------------------------------------===//
+#ifndef RTSAN_CHECK
+#error "Define RTSAN_CHECK prior to including this file!"
+#endif
+
+// RTSAN_CHECK(Name, SummaryKind)
+// SummaryKind should be a string literal.
+
+RTSAN_CHECK(FunctionCall, "function-call")
+RTSAN_CHECK(Interceptor, "interceptor")
```

Comment:
Do we like the names of these? What else would we call them

FunctionCall is "does this call exist anywhere in the stack"


Interceptor right now disables checks at the lowest level for a blocking function or an intercepted call by name (`malloc` or `my_call_with_attr_blocking`)

* Do we want to have two different disabling for each of them?
    * If so, what are their names and strings?
* Do we even want people to be able to disable [[blocking]] calls?

Reply:
Yeah good questions. Here's my initial thoughts:

_Q: Should we have different disabling keys for blocking, intercepted and stack functions?_

A:  I believe so, yes, for all three. I like to think we can improve on the names to make them more immediately understandable, even to people without a mental model of how the sanitizer works, but I'm struggling with `interceptor`. Here's my first iteration on the naming, but I don't think they're perfect at all, so keen to iterate more on them with you:

- "improve" `function-call` to something more like `call-stack-match`
- `interceptor` -> `intercepted`
- add `blocking`

_Q: Do we even want people to suppress `[[blocking]]` calls?

A: I think it could still be helpful, yes - I can imagine a situation where you're using a third-party library or re-using multi-threaded code in a single-threaded context (an noncontended spinlock here would always immediately grab a lock - the same reasoning here as it would apply to a normal mutex imho). I also worry that not suppressing them could potentially even be more confusing than otherwise. 

Another question for you:

_Q: Should we allow regex for `intercepted` and `blocking`? I can imagine a user might want to just suppress all `pthread_.*`, for example..._

Reply:
> Q: Should we have different disabling keys for blocking, intercepted and stack functions?
> A: I believe so, yes, for all three.

Sounds good, I'm in favor of this. I think the only reason to avoid any of it would be implementation/runtime complexity, but both of those are very tame. I think more atomic is better

> "improve" function-call to something more like call-stack-match
> interceptor -> intercepted
> add blocking

I like these generally, I'll give it some thought and we can continue hashing it out on the end PR

> _Q: Do we even want people to suppress [[blocking]] calls?
> A: I think it could still be helpful, yes

👍 

> Another question for you:
> Q: Should we allow regex for intercepted and blocking? I can imagine a user might want to just suppress all pthread_.*, for example...

Perhaps, but I would encourage punting on this until later on. The regex engine definitely will slow down the string matching. Before we go deeper with that I'd want to let the "basic" version prove to be inadequate for folks.

Basically "can we prove that we need it" before we implement something more complex. I think we are providing a lot of flexibility to the users already

Reply:
One correction here (of myself):

> Another question for you:
> Q: Should we allow regex for intercepted and blocking? I can imagine a user might want to just suppress all pthread_.*, for example...

So by default (and without me knowing) we ARE using the regex engine for function matches. This is the function we use:

```cpp
bool SuppressionContext::Match(const char *str, const char *type,
                               Suppression **s) {
  can_parse_ = false;
  if (!HasSuppressionType(type))
    return false;
  for (uptr i = 0; i < suppressions_.size(); i++) {
    Suppression &cur = suppressions_[i];
    if (0 == internal_strcmp(cur.type, type) && TemplateMatch(cur.templ, str)) {
      *s = &cur;
      return true;
    }
  }
  return false;
}

```

So early out for strcmp, then TemplateMatch, which is "regex light". It looks like it has a runtime of O(N), so it isn't more costly than just strcmp. We do both strcmp and TemplateMatch per function specified per type, so O(N_strlen*M_num_interceptor_suppressions).

Something to be aware of, that I don't think we will be able to get around - 

If the user is using suppressions, the more suppressions they use the more that run-time is affected. We need to be clear that using a ScopedDisabler is the first tool they should reach for if at all possible, which has an O(1) runtime cost. 

The hierarchy of speed:
* ScopedDisabler (easy, early out, never compare strings, never unwind the stack) - O(1)
* Function name suppressions, blocking and interceptors - O(N_strlen * M_suppressions_in_type)
* Stack suppressions - O(N_strlen * M_suppressions_in_type * L_stack_frame_count)

(I think)